mage_ai/data_cleaner/cleaning_rules/remove_outliers.py

```diff
@@ -1,36 +1,118 @@
+        1. If the data is more than 20 dimensional, reduce to 20 dimensions using principal component analysis
+        2. If there are 2 or more dimensions but less than or equal to 5 dimensions, use Local Outlier Factor method (LOF) to calculate
+           outliers. This is faster for smaller data size - theoretically, runtime of LOF grows quadratically with input
+        3. Otherwise, use Isolation Forests
+
+        Args:
+            df (DataFrame): Dataframe consisting of the numerical columns to perform outlier checks on
+            ndim (int): Number of dimensions for each example in dataframe
 
-            wrapped_c = wrap_column_name(c)
-            suggestions.append(
-                self._build_transformer_action_suggestion(
-                    REMOVE_OUTLIERS_TITLE,
-                    f'Remove {outlier_count} outlier(s) to reduce the amount of noise in the data.',
-                    ActionType.FILTER,
-                    action_arguments=[c],
-                    action_code=f'{wrapped_c} <= {upper} and {wrapped_c} >= {lower}',
-                    axis=Axis.ROW,
-                )
+        Returns:
+            Union[Dict, None]: If a suggestion is made, returns the suggestion dictionary. Else returns `None`
+        """
+        data = df.to_numpy()
+        count = data.shape[0]
+        if count <= 1:
+            return None
+        if ndim > 20:
+            data = self.pca_transformer.fit_transform(data)
+        if ndim <= 5:
+            if count < 10:
+                n_neighbors = 2
+            elif count < 500:
+                n_neighbors = count // 10 + 1
+            else:
+                n_neighbors = 20
+            outlier_algorithm = LocalOutlierFactor(n_neighbors=n_neighbors, n_jobs=-1)
+            outlier_algorithm.fit(data)
+            outlier_scores = outlier_algorithm.negative_outlier_factor_
+            outlier_mask = outlier_scores <= LOF_ANOMALY_SCORE_THRESHOLD
+        else:
+            outlier_algorithm = IsolationForest(n_estimators=75, n_jobs=-1, random_state=42)
+            outlier_algorithm.fit(data)
+            outlier_scores = outlier_algorithm.decision_function(data)
+            outlier_mask = outlier_scores <= ITREE_ANOMALY_SCORE_THRESHOLD
+        outlier_count = outlier_mask.sum()
+        if outlier_count > 0:
+            return self._build_transformer_action_suggestion(
+                REMOVE_OUTLIERS_TITLE,
+                f'Remove {outlier_count} outlier(s) to reduce the amount of noise in the data.',
+                ActionType.REMOVE,
+                action_options=dict(rows=df.index[outlier_mask].tolist()),
+                axis=Axis.ROW,
             )
+        return None
+
+    def evaluate(self):
+        suggestions = []
+        ndim = len(self.numeric_columns)
+        suggestion = None
+        if len(self.numeric_df) > 0:
+            if ndim == 1:
+                suggestion = self.one_dim_outlier_check(self.numeric_columns[0])
```

Comment:
we still want to provide one dimensional outlier check for all numeric columns.

Reply:
Will we make suggestions for both multidimensional outliers and univariate outliers as well?

Reply:
Yes. We'll provide both suggestions.

Reply:
Fixed, now both suggestions should be generated and suggestions for both update dynamically as outliers are removed